File: Python/9-DynamicProgramming(DP)/Problems/2-HouseRobber/HouseRobber.py
def rob_recursive(nums):
    """
    Recursive solution to the House Robber problem.

    This is the most basic approach, directly implementing the problem's
    recursive nature.  However, it's highly inefficient due to overlapping
    subproblems.

    Args:
        nums: A list of integers representing the amount of money in each house.

    Returns:
        The maximum amount of money the robber can steal.
    """
    def _rob(i):
        if i < 0:
            return 0
        if i == 0:
            return nums[0]
        # Option 1: Rob the current house (i) and skip the previous (i-1)
        rob_i = nums[i] + _rob(i - 2)
        # Option 2: Skip the current house (i) and consider the previous (i-1)
        skip_i = _rob(i - 1)
        return max(rob_i, skip_i)

    return _rob(len(nums) - 1)

def rob_memoization(nums):
    """
    Memoization (top-down dynamic programming) solution to the House Robber problem.

    This approach improves upon the recursive solution by storing the results of
    subproblems in a memo (dictionary).  This prevents redundant calculations.

    Args:
        nums: A list of integers representing the amount of money in each house.

    Returns:
        The maximum amount of money the robber can steal.
    """
    memo = {}  # Use a dictionary for memoization

    def _rob(i):
        if i < 0:
            return 0
        if i == 0:
            return nums[0]
        if i in memo:  # Check if the result is already memoized
            return memo[i]
        rob_i = nums[i] + _rob(i - 2)
        skip_i = _rob(i - 1)
        memo[i] = max(rob_i, skip_i)  # Store the result in the memo
        return memo[i]

    return _rob(len(nums) - 1)


def rob_tabulation(nums):
    """
    Tabulation (bottom-up dynamic programming) solution to the House Robber problem.

    This approach iteratively builds a table (dp array) of solutions to subproblems,
    starting from the base cases and working towards the final solution.

    Args:
        nums: A list of integers representing the amount of money in each house.

    Returns:
        The maximum amount of money the robber can steal.
    """
    n = len(nums)
    if n == 0:
        return 0
    if n == 1:
        return nums[0]

    dp = [0] * n  # dp[i] stores the max loot up to house i
    dp[0] = nums[0]
    dp[1] = max(nums[0], nums[1])  # Handle the case for the first two houses

    for i in range(2, n):
        #  Same logic as memoization, but using the dp table
        dp[i] = max(nums[i] + dp[i - 2], dp[i - 1])
    return dp[n - 1]

def rob_tabulation_optimized(nums):
    """
    Optimized tabulation solution to the House Robber problem.

    This is a space-optimized version of the tabulation approach.  Instead of
    storing the results for all subproblems, it only keeps track of the
    previous two maximum amounts, reducing the space complexity from O(n) to O(1).

    Args:
        nums: A list of integers representing the amount of money in each house.

    Returns:
        The maximum amount of money the robber can steal.
    """
    n = len(nums)
    if n == 0:
        return 0
    if n == 1:
        return nums[0]

    #  prev1 stores max loot up to i-1, prev2 stores max loot up to i-2
    prev2 = nums[0]
    prev1 = max(nums[0], nums[1])

    for i in range(2, n):
        # Calculate current max loot
        current = max(nums[i] + prev2, prev1)
        # Update prev2 and prev1 for the next iteration
        prev2 = prev1
        prev1 = current
    return prev1

def rob_iterative(nums):
    """
    Iterative solution that avoids using extra space (O(1)).

    This approach is similar to the optimized tabulation method but written in a more
    straightforward iterative style. It explicitly tracks the maximum loot obtainable
    at the previous two houses and updates them as it iterates through the houses.

    Args:
        nums (list[int]): A list of integers representing the amount of money in each house.

    Returns:
        int: The maximum amount of money the robber can steal.
    """
    n = len(nums)
    if n == 0:
        return 0
    if n == 1:
        return nums[0]

    # Initialize variables to store the maximum loot at the previous two houses.
    # `rob_prev` stores the maximum loot obtainable up to the (i-1)-th house.
    # `not_rob_prev` stores the maximum loot obtainable up to the (i-2)-th house.
    rob_prev = 0
    not_rob_prev = 0

    # Iterate through the houses.
    for i in range(n):
        # Calculate the maximum loot if we rob the current house.
        rob_current = not_rob_prev + nums[i]
        # Calculate the maximum loot if we don't rob the current house.
        not_rob_current = max(rob_prev, not_rob_prev)

        # Update `rob_prev` and `not_rob_prev` for the next iteration.
        rob_prev = rob_current
        not_rob_prev = not_rob_current

    # The maximum loot is the greater of robbing the last house or not robbing it.
    return max(rob_prev, not_rob_prev)
# Example usage:
if __name__ == "__main__":
    houses = [2, 7, 9, 3, 1]
    print("Houses:", houses)

    print("\nRecursive Solution:", rob_recursive(houses))
    print("Memoization Solution:", rob_memoization(houses))
    print("Tabulation Solution:", rob_tabulation(houses))
    print("Tabulation Optimized Solution:", rob_tabulation_optimized(houses))
    print("Iterative Solution:", rob_iterative(houses))

    houses2 = [1, 2, 3, 1]
    print("\nHouses:", houses2)
    print("Recursive Solution:", rob_recursive(houses2))
    print("Memoization Solution:", rob_memoization(houses2))
    print("Tabulation Solution:", rob_tabulation(houses2))
    print("Tabulation Optimized Solution:", rob_tabulation_optimized(houses2))
    print("Iterative Solution:", rob_iterative(houses2))

    houses3 = [2, 1]
    print("\nHouses:", houses3)
    print("Recursive Solution:", rob_recursive(houses3))
    print("Memoization Solution:", rob_memoization(houses3))
    print("Tabulation Solution:", rob_tabulation(houses3))
    print("Tabulation Optimized Solution:", rob_tabulation_optimized(houses3))
    print("Iterative Solution:", rob_iterative(houses3))

    houses4 = [1]
    print("\nHouses:", houses4)
    print("Recursive Solution:", rob_recursive(houses4))
    print("Memoization Solution:", rob_memoization(houses4))
    print("Tabulation Solution:", rob_tabulation(houses4))
    print("Tabulation Optimized Solution:", rob_tabulation_optimized(houses4))
    print("Iterative Solution:", rob_iterative(houses4))

    houses5 = []
    print("\nHouses:", houses5)
    print("Recursive Solution:", rob_recursive(houses5))
    print("Memoization Solution:", rob_memoization(houses5))
    print("Tabulation Solution:", rob_tabulation(houses5))
    print("Tabulation Optimized Solution:", rob_tabulation_optimized(houses5))
    print("Iterative Solution:", rob_iterative(houses5))

File: C++/9-DynamicProgramming(DP)/Problems/2-HouseRobber/HouseRobber.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric> // For std::accumulate

using namespace std;

// 1. Top-Down Recursion with Memoization
//    - Time Complexity: O(n), where n is the number of houses.
//    - Space Complexity: O(n) for the memoization table and recursion stack.
int robTopDown(const vector<int>& nums, int i, vector<int>& memo) {
    if (i < 0) {
        return 0;
    }
    if (memo[i] != -1) {
        return memo[i];
    }

    // Option 1: Rob the current house i and skip the previous house (i-1).
    int robHouse = nums[i] + robTopDown(nums, i - 2, memo);
    // Option 2: Skip the current house i and consider the maximum from previous houses.
    int skipHouse = robTopDown(nums, i - 1, memo);

    memo[i] = max(robHouse, skipHouse); // Store the maximum for future use.
    return memo[i];
}

int rob_top_down(vector<int>& nums) {
    if (nums.empty()) return 0;
    vector<int> memo(nums.size(), -1); // Initialize memoization table with -1.
    return robTopDown(nums, nums.size() - 1, memo);
}

// 2. Bottom-Up Dynamic Programming (Tabulation) - Optimized
//    - Time Complexity: O(n)
//    - Space Complexity: O(1) - Constant space
int robBottomUpOptimized(vector<int>& nums) {
    if (nums.empty()) return 0;
    if (nums.size() == 1) return nums[0];

    int prev2 = 0;       // Represents dp[i-2]
    int prev1 = 0;       // Represents dp[i-1]
    int current;

    for (int i = 0; i < nums.size(); ++i) {
        // Calculate the maximum amount that can be robbed up to house i.
        current = max(prev1, prev2 + nums[i]); // Either rob house i or don't.
        prev2 = prev1;  // Update prev2 (dp[i-2])
        prev1 = current; // Update prev1 (dp[i-1])
    }
    return prev1; // The result is stored in prev1 after the loop.
}

// 3. Bottom-Up Dynamic Programming (Tabulation) - Array
//    - Time Complexity: O(n)
//    - Space Complexity: O(n)
int robBottomUpArray(vector<int>& nums) {
    if (nums.empty()) return 0;
    if (nums.size() == 1) return nums[0];

    vector<int> dp(nums.size()); // dp[i] stores the max loot up to house i.
    dp[0] = nums[0];            // Base case: Only one house.
    dp[1] = max(nums[0], nums[1]); // Base case: Two houses - rob the one with max value.

    for (int i = 2; i < nums.size(); ++i) {
        // dp[i] is the maximum of:
        // 1. Robbing the current house i and adding it to the loot from i-2.
        // 2. Not robbing house i and taking the best loot from i-1.
        dp[i] = max(dp[i - 1], dp[i - 2] + nums[i]);
    }
    return dp[nums.size() - 1]; // Result: max loot up to the last house.
}

// 4.  Recursion (Plain Recursion - Exceeds Time Limit for larger inputs)
//     - Time Complexity: O(2^n) - Exponential (very slow for larger n)
//     - Space Complexity: O(n) -  Due to the recursion call stack.
int robRecursive(const vector<int>& nums, int i) {
    if (i < 0) {
        return 0;
    }
    // Option 1: Rob current house and skip the previous
    int robHouse = nums[i] + robRecursive(nums, i - 2);
    // Option 2: Skip current house
    int skipHouse = robRecursive(nums, i - 1);
    return max(robHouse, skipHouse);
}

int rob_recursive(vector<int>& nums) {
    if (nums.empty()) return 0;
    return robRecursive(nums, nums.size() - 1);
}

// 5. Iterative with two variables, similar to optimized bottom up, but written slightly differently
int robIterative(vector<int>& nums) {
    if (nums.empty()) return 0;
    if (nums.size() == 1) return nums[0];

    int rob = nums[0]; // Max loot including the current house
    int notRob = 0;    // Max loot excluding the current house

    for (int i = 1; i < nums.size(); ++i) {
        int newRob = max(notRob + nums[i], rob); // Current house robbed
        int newNotRob = max(rob, notRob);    // Current house not robbed

        rob = newRob;
        notRob = newNotRob;
    }
    return rob;
}

int main() {
    vector<int> houses = {2, 7, 9, 3, 1}; // Example house values

    cout << "House Robber Problem" << endl;
    cout << "House values: ";
    for (int val : houses) {
        cout << val << " ";
    }
    cout << endl;

    cout << "\n1. Top-Down Recursion with Memoization: " << rob_top_down(houses) << endl;
    cout << "2. Bottom-Up Dynamic Programming (Optimized): " << robBottomUpOptimized(houses) << endl;
    cout << "3. Bottom-Up Dynamic Programming (Array): " << robBottomUpArray(houses) << endl;
    cout << "4. Plain Recursion: " << rob_recursive(houses) << endl;
    cout << "5. Iterative with two variables: " << robIterative(houses) << endl;

    return 0;
}
